File: scripts/optimize-images.js
#!/usr/bin/env node
/**
 * Node.js 이미지 최적화 스크립트
 * - 모든 이미지를 WebP로 변환
 * - 파일 크기 80% 압축
 * - 최대 너비 1920px로 리사이징
 * - 원본 파일은 backup 폴더로 이동
 */

const fs = require('fs');
const path = require('path');
const sharp = require('sharp');

// 백업 디렉토리 생성
function createBackupDir() {
    const backupDir = path.join('images', 'backup');
    if (!fs.existsSync(backupDir)) {
        fs.mkdirSync(backupDir, { recursive: true });
    }
    return backupDir;
}

// 이미지 최적화 및 WebP 변환
async function optimizeImage(inputPath, outputPath, quality = 80, maxWidth = 1920) {
    try {
        const image = sharp(inputPath);
        const metadata = await image.metadata();

        // 리사이징 (너비 기준)
        let resizedImage = image;
        if (metadata.width > maxWidth) {
            const ratio = maxWidth / metadata.width;
            const newHeight = Math.round(metadata.height * ratio);
            resizedImage = image.resize(maxWidth, newHeight);
        }

        // WebP로 변환 및 저장
        await resizedImage
            .webp({ quality: quality, effort: 6 })
            .toFile(outputPath);

        // 파일 크기 비교
        const originalStats = fs.statSync(inputPath);
        const newStats = fs.statSync(outputPath);
        const compressionRatio = (1 - newStats.size / originalStats.size) * 100;

        console.log(`✅ ${path.basename(inputPath)} → ${path.basename(outputPath)}`);
        console.log(`   크기: ${Math.round(originalStats.size / 1024)}KB → ${Math.round(newStats.size / 1024)}KB (${compressionRatio.toFixed(1)}% 압축)`);

        return { success: true, originalSize: originalStats.size, newSize: newStats.size };
    } catch (error) {
        console.log(`❌ ${inputPath} 처리 실패: ${error.message}`);
        return { success: false, originalSize: 0, newSize: 0 };
    }
}

// 메인 실행 함수
async function main() {
    console.log('🖼️  이미지 최적화 시작...');
    console.log('='.repeat(50));

    // Sharp 라이브러리 확인
    try {
        require('sharp');
    } catch (error) {
        console.log('❌ Sharp 라이브러리가 설치되지 않았습니다.');
        console.log('💡 다음 명령으로 설치해주세요: npm install sharp');
        process.exit(1);
    }

    // 백업 디렉토리 생성
    const backupDir = createBackupDir();

    // 이미지 확장자 목록
    const imageExtensions = ['.jpg', '.jpeg', '.png', '.bmp', '.tiff', '.tif', '.webp'];

    // images 폴더의 모든 파일 처리
    const imagesDir = 'images';
    let optimizedCount = 0;
    let totalOriginalSize = 0;
    let totalNewSize = 0;

    try {
        const files = fs.readdirSync(imagesDir);

        for (const filename of files) {
            const filePath = path.join(imagesDir, filename);
            const stats = fs.statSync(filePath);

            // 파일인지 확인하고 이미지 확장자인지 체크
            if (stats.isFile()) {
                const fileExt = path.extname(filename).toLowerCase();

                if (imageExtensions.includes(fileExt)) {
                    const baseName = path.basename(filename, fileExt);
                    const outputFilename = `${baseName}.webp`;
                    const outputPath = path.join(imagesDir, outputFilename);

                    // 이미지 최적화
                    const result = await optimizeImage(filePath, outputPath);

                    if (result.success) {
                        totalOriginalSize += result.originalSize;
                        totalNewSize += result.newSize;

                        // 원본 파일을 백업 폴더로 이동 (이미 WebP가 아닌 경우만)
                        if (fileExt !== '.webp') {
                            const backupPath = path.join(backupDir, filename);
                            fs.renameSync(filePath, backupPath);
                            console.log(`   원본 백업: ${backupPath}`);
                        }

                        optimizedCount++;
                        console.log();
                    }
                }
            }
        }

        // 최종 결과 출력
        console.log('='.repeat(50));
        console.log('🎉 최적화 완료!');
        console.log(`📊 처리된 이미지: ${optimizedCount}개`);
        console.log(`💾 전체 크기: ${Math.round(totalOriginalSize / 1024)}KB → ${Math.round(totalNewSize / 1024)}KB`);

        if (totalOriginalSize > 0) {
            const totalCompression = (1 - totalNewSize / totalOriginalSize) * 100;
            console.log(`🗜️  전체 압축률: ${totalCompression.toFixed(1)}%`);
            console.log(`⚡ 예상 로딩 속도 개선: ${Math.round(totalCompression)}%`);
        }

    } catch (error) {
        console.log(`❌ 폴더 읽기 실패: ${error.message}`);
    }
}

// 스크립트 실행
if (require.main === module) {
    main().catch(console.error);
} 

File: scripts/optimize-images.py
#!/usr/bin/env python3
"""
이미지 최적화 스크립트
- 모든 이미지를 WebP로 변환
- 파일 크기 80% 압축
- 최대 너비 1920px로 리사이징
- 원본 파일은 백업 폴더로 이동
"""

import os
import shutil
from PIL import Image
import pillow_heif

# HEIF 지원 등록
pillow_heif.register_heif_opener()

def create_backup_dir():
    """백업 디렉토리 생성"""
    backup_dir = "images/backup"
    if not os.path.exists(backup_dir):
        os.makedirs(backup_dir)
    return backup_dir

def optimize_image(input_path, output_path, quality=80, max_width=1920):
    """이미지 최적화 및 WebP 변환"""
    try:
        with Image.open(input_path) as img:
            # RGBA 모드인 경우 RGB로 변환 (WebP 호환성)
            if img.mode in ('RGBA', 'LA', 'P'):
                # 투명 배경을 흰색으로 변환
                background = Image.new('RGB', img.size, (255, 255, 255))
                if img.mode == 'P':
                    img = img.convert('RGBA')
                background.paste(img, mask=img.split()[-1] if img.mode in ('RGBA', 'LA') else None)
                img = background
            elif img.mode != 'RGB':
                img = img.convert('RGB')
            
            # 리사이징 (너비 기준)
            if img.width > max_width:
                ratio = max_width / img.width
                new_height = int(img.height * ratio)
                img = img.resize((max_width, new_height), Image.Resampling.LANCZOS)
            
            # WebP로 저장
            img.save(output_path, 'WebP', quality=quality, optimize=True)
            
            # 파일 크기 비교
            original_size = os.path.getsize(input_path)
            new_size = os.path.getsize(output_path)
            compression_ratio = (1 - new_size / original_size) * 100
            
            print(f"✅ {os.path.basename(input_path)} → {os.path.basename(output_path)}")
            print(f"   크기: {original_size // 1024}KB → {new_size // 1024}KB ({compression_ratio:.1f}% 압축)")
            
            return True
            
    except Exception as e:
        print(f"❌ {input_path} 처리 실패: {e}")
        return False

def main():
    """메인 실행 함수"""
    print("🖼️  이미지 최적화 시작...")
    print("=" * 50)
    
    # 백업 디렉토리 생성
    backup_dir = create_backup_dir()
    
    # 이미지 확장자 목록
    image_extensions = {'.jpg', '.jpeg', '.png', '.bmp', '.tiff', '.tif', '.webp'}
    
    # images 폴더의 모든 파일 처리
    images_dir = "images"
    optimized_count = 0
    total_original_size = 0
    total_new_size = 0
    
    for filename in os.listdir(images_dir):
        file_path = os.path.join(images_dir, filename)
        
        # 파일인지 확인하고 이미지 확장자인지 체크
        if os.path.isfile(file_path):
            file_ext = os.path.splitext(filename)[1].lower()
            
            if file_ext in image_extensions:
                # 원본 파일 크기 계산
                original_size = os.path.getsize(file_path)
                total_original_size += original_size
                
                # 출력 파일명 (WebP로 변환)
                base_name = os.path.splitext(filename)[0]
                output_filename = f"{base_name}.webp"
                output_path = os.path.join(images_dir, output_filename)
                
                # 이미지 최적화
                if optimize_image(file_path, output_path):
                    # 최적화된 파일 크기 계산
                    new_size = os.path.getsize(output_path)
                    total_new_size += new_size
                    
                    # 원본 파일을 백업 폴더로 이동 (이미 WebP가 아닌 경우만)
                    if file_ext != '.webp':
                        backup_path = os.path.join(backup_dir, filename)
                        shutil.move(file_path, backup_path)
                        print(f"   원본 백업: {backup_path}")
                    
                    optimized_count += 1
                    print()
    
    # 최종 결과 출력
    print("=" * 50)
    print(f"🎉 최적화 완료!")
    print(f"📊 처리된 이미지: {optimized_count}개")
    print(f"💾 전체 크기: {total_original_size // 1024}KB → {total_new_size // 1024}KB")
    
    if total_original_size > 0:
        total_compression = (1 - total_new_size / total_original_size) * 100
        print(f"🗜️  전체 압축률: {total_compression:.1f}%")
        print(f"⚡ 예상 로딩 속도 개선: {total_compression:.0f}%")

if __name__ == "__main__":
    main() 